src/game_api/thread_utils.hpp

```diff
@@ -1,57 +1,132 @@
 #pragma once
 
-#include <Windows.h> // for ResumeThread, SuspendThread, HANDLE
-#include <cstddef>   // for size_t
-#include <cstdint>   // for int64_t
+#include <cstddef>  // for size_t
+#include <cstdint>  // for int64_t
+#include <stdlib.h> // for free
 
-HANDLE get_main_thread();
-size_t heap_base();
-size_t local_heap_base();
+struct PRNG;
+struct StateMemory;
+struct LevelGenSystem;
+struct LiquidPhysics;
 
-// Used for objects that are allocated with the game's custom allocator
-template <typename T>
-class OnHeapPointer
+struct HeapBase
 {
-    int64_t ptr_;
+    static HeapBase get(uint8_t slot);
+    static HeapBase get_main();
+    // can be NULL
+    static HeapBase get_local();
+    // fallback to main if can't get local
+    static HeapBase get_local_safe()

```

Comment:
maybe instead of having `get_local` and `get_local_safe` i should flip them, so `get_local_unsafe` and `get_local` ?

Reply:
or maybe even go minimalist approach, 'get` (which will be the local safe) and 'get_main`

Reply:
I'm pretty sure the issue with local state being null is the way overlunky injects and runs some code in the thread that's injecting to it. If we make it to run code only on the spelunky thread, it shouldn't be null. iirc I got something to work in the third online changes PR, which isn't even open yet

Reply:
for now i will just build in nullpointer check and make `get` and `get_main`

Reply:
I think this is it
https://github.com/estebanfer/overlunky/commit/fc8b2c190a7558066448fefa89451a9c042efdc0#diff-20c19a6e71839b8083a6988dfb4bfb30a6598122ad4cffb83de6ffd418a176e7

Reply:
ok cool, you want me to move that here?
but also, i needed to add this cuz it was crashing (don't remember if it was only PL maybe?)
https://github.com/spelunky-fyi/overlunky/blob/main/src/game_api/script/lua_vm.cpp#L147-L152

Reply:
> ok cool, you want me to move that here?

if you want, iirc I just made that change quick and without thinking much, and it worked to make the local state valid at overlunky start, and I never got any issues elsewhere. I'm not sure if how it's coded it's really the best way to do it, but I guess it should be something like that

Reply:
> but also, i needed to add this cuz it was crashing (don't remember if it was only PL maybe?)

hmm, I never got found any nullptr return using the local ptr with PL, not that I remember